pkg/controller/job/job_controller.go

```diff
@@ -1001,7 +1016,7 @@ func (jm *Controller) trackJobStatusAndRemoveFinalizers(ctx context.Context, job
 			continue
 		}
 		considerPodFailed := isPodFailed(pod, job)
-		if podutil.IsPodTerminal(pod) || considerPodFailed || finishedCond != nil || job.DeletionTimestamp != nil {
+		if canRemoveFinalizer(logger, job, finishedCond, pod, considerPodFailed, podsWithDelayedDeletionPerIndex) {
```

Comment:
maybe we need to account for `podsWithDelayedDeletionPerIndex` when doing `enactFinishedJob`?

Reply:
We already account for them indirectly by checking the uncounted pods are empty: https://github.com/kubernetes/kubernetes/blob/98c6e25c37128c10e238bf24a31356c9b4540d55/pkg/controller/job/job_controller.go#L1233C20-L1233C29. We only transfer pods from uncounted to Failed counter when they have finalizers removed in `cleanUncountedPodsWithoutFinalizers`. So, the pods with delayed finalizer removal will already delay `enactJobFinished`